### PR TITLE
Add support for signing AWS event stream messages

### PR DESCRIPTION
**Pull request by dmorn**
This is the code we're currently using for signing event stream messages in conjunction with
our fork of https://github.com/kim-company/event_stream.

I do not expect this to be ready to be merged, it was a quick-and-dirty feature addition, but
I would like to open the PR anyway to give an idea of a working implementation.

If you consider this a good contribution to the library, I'll be happy to help sorting this out.

### CODE REVIEW

**Comment by jonatanklosko** [comment]
Hey @dmorn, adding the function sounds good to me, FTR [here](https://github.com/aws/aws-sdk-ruby/blob/fcb041040567084f89d307c0814e3d6fb5a3afbf/gems/aws-sigv4/lib/aws-sigv4/signer.rb#L315) is a reference Ruby implementation.

If we can reasonably add headers in the function, then I would make it `sign_v4_event` that returns `{headers, signature}`. Otherwise it can be something like `sign_v4_signature_for_event` that returns just the signature.

A test would be very welcome as well!

Nitpick: please drop formatting changes from the PR :)

**Comment by onno-vos-dev** [comment]
Will review later @dmorn any chance you can back out some of the re-formatting? It makes the diff somewhat harder to follow as to what actually changed and what has been re-formatted. Alternative, put the re-formatting in it's own explicit commit.

**Comment by dmorn** [comment]
Hi @jonatanklosko and @onno-vos-dev! Yes, I'll definitely revert the formatting changes. I'm going to be AFK for 2 weeks now but I'll sort this out as soon as I come back!

**Comment by jonatanklosko** [comment]
@dmorn friendly ping in case you still want to work on this :) There is definitely no rush though!

**Comment by dmorn** [comment]
Sorry @jonatanklosko I completely forgot about it, thanks for the ping 😱 I'll update this PR soon ✌️

**Comment by onno-vos-dev** [comment]
@dmorn LGTM and thanks for backing out the reformatting changes. I was gonna ask to add some tests but just realized we don't have any tests for aws_signature 😞 😢 (It's typically tested through `aws-erlang | aws_elixir` I'll follow up on this myself 👍

**Comment by jonatanklosko** [comment]
@onno-vos-dev we do have tests, they are in `aws_signature.erl` :)

@dmorn if you could add a test that would be great. Also, please add some documentation to the function :)

**Comment by onno-vos-dev** [comment]
@jonatanklosko ofcourse they are 🤦

**Comment by dmorn** [comment]
@jonatanklosko @onno-vos-dev it may be it. I would like to point to some official docs but the only (incomplete) piece of information I can find is in https://docs.aws.amazon.com/transcribe/latest/dg/streaming-setting-up.html. When I originally worked on this I found a more comprehensive document that I cannot find anymore 🫠

**Comment by dmorn** [comment]
It would also be useful to point to the EventStream library, but unfortunately https://github.com/VoiceLayer/event_stream cannot be used as it is right now as it won't encode the headers correctly; for that, users should use our fork... I'm pinging the maintainers

File: src/aws_signature.erl [hunk 1]
```diff
@@ -94,6 +94,54 @@ sign_v4(AccessKeyID, SecretAccessKey, Region, Service, DateTime, Method, URL, He
...
+%% Headers of a sigv4 signed event message only contains 2 headers
+%%   * ':chunk-signature'
+%%     * computed signature of the event, binary string, 'bytes' type
+%%   * ':date'
+%%     * millisecond since epoch, 'timestamp' type
+%%
+%% `PriorSignature' for the first message is the base16 encoded signv4
+%% of the request used to open a connection with the target service.
+%%
+%% `HeadersSing' are the headers of the inner packet, encoded using the
```
Comment (jonatanklosko) [inline src/aws_signature.erl:109]:
```suggestion
%% `HeadersString' are the headers of the inner packet, encoded using the
```

File: src/aws_signature.erl [hunk 2]
```diff
@@ -94,6 +94,54 @@ sign_v4(AccessKeyID, SecretAccessKey, Region, Service, DateTime, Method, URL, He
 
     add_authorization_header(FinalHeaders, Authorization).
 
+%% @doc Signs an AWS Event Stream message and returns the headers and
+%% prior signature used for next event signing.
```
Comment (jonatanklosko) [inline src/aws_signature.erl:98]: It's prior only in the context of next signing :)

```suggestion
%% signature used for next event signing.
```

File: src/aws_signature.erl [hunk 3]
```diff
@@ -94,6 +94,54 @@ sign_v4(AccessKeyID, SecretAccessKey, Region, Service, DateTime, Method, URL, He
...
     add_authorization_header(FinalHeaders, Authorization).
 
+%% @doc Signs an AWS Event Stream message and returns the headers and
+%% signature used for next event signing.
+%%
+%% Headers of a sigv4 signed event message only contains 2 headers
+%%   * ':chunk-signature'
+%%     * computed signature of the event, binary string, 'bytes' type
+%%   * ':date'
+%%     * millisecond since epoch, 'timestamp' type
```
Comment (jonatanklosko) [inline src/aws_signature.erl:104]: A markdown list won't work here, we need to do same as here:

https://github.com/aws-beam/aws_signature/blob/a40e674f9681f70a6bc5d75035dd3edf7372c8c5/src/aws_signature.erl#L177-L184

Comment (dmorn) [reply src/aws_signature.erl]: I updated the docs!

File: src/aws_signature.erl [hunk 4]
```diff
@@ -94,6 +94,58 @@ sign_v4(AccessKeyID, SecretAccessKey, Region, Service, DateTime, Method, URL, He
...
+%% signature used for next event signing.
+%%
+%% Headers of a sigv4 signed event message only contains 2 headers
+%% <dl>
+%% <dt>`:chunk-signature'</dt>
+%% <dd>
+%% computed signature of the event, binary string, `bytes' type
+%% </dd>
+%% <dt>`:date'</dt>
+%% millisecond since epoch, `timestamp' type
```
Comment (jonatanklosko) [inline src/aws_signature.erl:107]:
```suggestion
%% <dd>
%% millisecond since epoch, `timestamp' type
%% </dd>
```

**Review by jonatanklosko — APPROVED** [review]

**Comment by dmorn** [comment]
@jonatanklosko  are we eventually merging this? 😉

**Comment by jonatanklosko** [comment]
Oh! @onno-vos-dev feel free to merge whenever you are happy :)

**Comment by onno-vos-dev** [comment]
Awesome! Thank you! <3 Merged, will tag shortly. Sorry for the delay :-)

**Pull request merged**